### DISCUSSION TITLE
How to handle next/navigation's redirect method when testing RSC using RTL + Jest?

### DISCUSSION THREAD

**Discussion by kdb13** [vercel/next.js#59061] [Help]
### Summary

I'm testing a RSC `page.tsx` using RTL and Jest. My component is calling the `next/navigation`'s `redirect()` based on some specific values in the `searchParams`. When I render the component inside the test, the test fails at redirect() method with the error `NEXT_REDIRECT`.

What is the best way to handle this scenario? If I try to mock the `redirect` method, then the component won't stop executing after the redirect method is called which results in an inconsistent state.

**Comment by IMAMBAKS** [comment]
Have you found any solution?

**Reply by avance-xealth** [reply]
Here's one way I've found to test the redirect. The idea is to set up `redirect()` to throw an error so that execution stops.

MyComponent.tsx
```typescript
import { redirect } from 'next/navigation'

export default function MyComponent() {
  redirect('/foo')
}
```

MyComponent.spec.tsx
```typescript
import { redirect } from 'next/navigation'
import MyComponent from './MyComponent'

class TestRedirect extends Error {
  readonly url: string
  constructor(url: string) {
    super()
    this.url = url
  }
}

jest.mock('next/navigation')
jest.mocked(redirect).mockImplementation((url: string) => { throw new TestRedirect(url) })

it('redirects', () => {
  let redirect: TestRedirect | undefined
  try {
    render(<MyComponent />)
  } catch (e) {
    redirect = e as TestRedirect
  }
  expect(redirect?.url).toBe('/foo')
})
```
